docs/question.md

```diff
@@ -447,3 +447,90 @@ For most [elements] there are four different ways of grading the student answer.
 4. Write an [external grader](externalGrading.md), though this is typically applied to more complex questions like coding.
 
 If a question has more than one of the above options, each of them overrides the one before it. Even if options 3 (custom grade function) or 4 (external grader) are used, then it can still be helpful to set a correct answer so that it is shown to students as a sample of what would be accepted. If there are multiple correct answers then it's probably a good idea to add a note with [`pl-answer-panel`](elements/#pl-answer-panel-element) that any correct answer would be accepted and this displayed answer is only an example.
+
+## Initializing parameters from outside the question
+
+The `data["params"]` dictionary normally used to pass parameters to `question.html` can be "seeded" from outside the question's files, allowing some portion of its randomization to be set by an assessment, course instance, or course configuration file. Such "`questionParams`" can change the behavior of a question without needing to alter its source code, so a question can be reused on different assessments but with a different configuration of parameters to alter its actual variants as needed.
+
+A `questionParams` is a JSON object that can be placed in three different PrairieLearn file types. When a question's `generate()` function is called, the `data["params"]` dictionary will be pre-filled with the key-value pairs defined by `questionParams` in these locations:
+
+1. `infoCourse.json`: parameters defined here will be passed to all questions in the course, including when they are previewed by an instructor without being attached to an assessment.
```

Comment:
I'm concerned that `questionParams` set in `infoCourse.json` will "pollute" the `params` of all questions. That is, every variant generated in the course will include the course `questionParams` in its stored `params`. For example, in the example below where course instructor names are set in the `infoCourseInstance.json` question params, I think that every variant for every question in that course instance would store the instructor names?

I'm not really worried about the storage, it's more that it will be messy when we display the params for variants and for most questions there are these spurious entries.

A couple of ideas off the top of my head to address this:

1. Use `questionOptions` rather than `questionParams` and merge into `data["options"]`. This is not stored on a per-variant basis. If a question wants to store the data, they can copy it from `options` into `params`.

2. Require questions to opt-in to particular question params by listing them in their `info.json` file. For example, a question would need to declare `"useQuestionParams": ["names"]` in its `info.json` file to have the `names` question param copied into its local `params`.

Reply:
Yes, I understand your concern. It's ultimately your team's call, both solutions sound easy enough for me to implement. Here are my thoughts:

1. This is nice for more clearly differentiating "params" (which currently only come from server.py, right?) from "options" (which are passed in from external sources). I think most use of "options" will be to set bounds/configurations on the variants that are generated (at least this is true in my head), and those don't really need to be visible.  As you say, we can copy options to params if we want them visible/stored.
2. This is also nice for documenting the ways a question can be configured.

Reply:
@nwalters512 thoughts about this?

Reply:
@nealterrell Matt and just chatted about this for an hour or so. We've agreed that something more like option 2 is the way we want to go, i.e. use params that are persisted as part of a variant (not options) and require questions to declare somehow that they will or can use a certain parameter. We'd like to spend a little bit longer talking about this, as the design will have implications for other things that are part of our longer-term vision for PL, and we want to make sure that we're at least foreseeing potential problems now. Sorry the complexity is snowballing here, but we want to be really deliberate with how we're designing this.

Matt is heading off to Learning@Scale for a few days so we probably won't be able to talk about this more til next week. In the meantime, here are some of the things we're thinking about:

- Someday, we'll allow folks to share questions between courses or institutions. How can we make sure we're guiding people towards writing sharable questions in the context of params, which may not be present in a course that a question is shared with? We're thinking about some way to declare defaults in `info.json`, and possibly making that mandatory.
  - The alternative is of course `.get('...', DEFAULT)` in Python, but if your course or course instance already has that value in place, it would be easy to forget that you should specify a default value as you'd never hit the case of a missing param during question development.
- In the same vein as defaults, having declarative information about params in `info.json` would allow us to do things like make an editor or preview UI to show which params will take effect, or to layer on some kind of schema checking, and so on. We probably want something more than just a list of param names, but how much more? Maybe `{"foo": {"default: 5}}`? Or `[{"name": "foo", "default": 5}]`? Maybe we just use [JSON Schema](https://json-schema.org/) directly?
- What do we actually want to call these things? Calling them just "params" inside a question might be a bit confusing, as they're different from the params that a question would write into `params` during `generate`. Maybe we call them "inherited params" or "external params"? @mwest1066 might be able to expand on this thought a bit.

Feel free to chime in with your thoughts about any/all of the above. I'll keep you posted with what Matt and I chat about next week.

Reply:
To chime in briefly about this, having the "params" field in the data dict copied into could cause confusion in designing the backend for a question. The easiest thing (IMO) would be to have a separate entry in the data dictionary (`questionParams`, for example) that only holds options taken from the `info.json` (similar to option 2). That way, we can add `questionParams` to the type signatures for the data dict, and it makes it easier to declare a custom dictionary for just those parameters. This seems like a really cool feature, and something that is a good stepping stone to question sharing across courses.

Reply:
I'm certainly not opposed to having these "external" params copied into a separate entry in `data`. What to call it is not so clear to me, however 😄. `questionParams` seems like a good name to use in `infoAssessment.json`, but I feel like it would be weird to have both `data['params']` and `data['questionParams']`.

And @nwalters512 and I really need to figure this out, we'd love to get this feature merged.

Reply:
`explicitParams`? `specParams`? `configuredParams`?

Reply:
Maybe `assignmentParams`? The whole point is that these parameters are fed in externally from the assignment configuration.

Reply:
The word "external" is coming up a lot, so I'd also suggest `externalParams`.

Reply:
I think I agree with @echuber2, the parameters can come from the course, instance, or assignment level, so `externalParams` seems more natural. @nwalters512 what do you think?

Reply:
I'm a big fan of `externalParams`!

Reply:
With respect to declaring that information in the `info.json`, it seems like having a dictionary of the form `{"foo": {"default": 5}}` as Nathan mentioned would be best, since dict keys have to be distinct (as opposed to the second list suggestion). So the entry `externalParams` in the `info.json` for a question would hold that dict, and then the course + assignment + whatever could override it as needed (with an error being thrown if a parameter is injected that isn't on the question's approved list). Handling this with the json as opposed to just in the Python backend, the error message will be thrown as soon as the assignment configuration including the question is created, as opposed to when the question is first opened.

Also I know this has been sitting around for a while, but @nealterrell let us know if you'd like any help implementing this, since our team has grown enough that we can offer some assistance if needed, and this would be a useful feature for us.

Reply:
@eliotwrobson I am able to work on the trivial stuff like renaming the JSON fields, but the plans for parameter schemas and validation probably exceeds my bandwidth at the moment.

Reply:
Understood, I'll get someone on that piece of this as soon as I can. The parameter validation shouldn't be _too_ difficult, since you already have the parameter injection set up. Will follow up later if we run into issues.